File: sdks/dotnet/sulfone-helium/Api/Core/QuestionRes.cs
using System.Text.Json.Serialization;
using Swashbuckle.AspNetCore.Annotations;

namespace sulfone_helium.Api.Core;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[SwaggerDiscriminator("type")]
[JsonDerivedType(typeof(ConfirmQuestionRes), typeDiscriminator: "confirm")]
[SwaggerSubType(typeof(ConfirmQuestionRes), DiscriminatorValue = "confirm")]
[JsonDerivedType(typeof(DateQuestionRes), typeDiscriminator: "date")]
[SwaggerSubType(typeof(DateQuestionRes), DiscriminatorValue = "date")]
[JsonDerivedType(typeof(CheckboxQuestionRes), typeDiscriminator: "checkbox")]
[SwaggerSubType(typeof(CheckboxQuestionRes), DiscriminatorValue = "checkbox")]
[JsonDerivedType(typeof(PasswordQuestionRes), typeDiscriminator: "password")]
[SwaggerSubType(typeof(PasswordQuestionRes), DiscriminatorValue = "password")]
[JsonDerivedType(typeof(TextQuestionRes), typeDiscriminator: "text")]
[SwaggerSubType(typeof(TextQuestionRes), DiscriminatorValue = "text")]
[JsonDerivedType(typeof(SelectQuestionRes), typeDiscriminator: "select")]
[SwaggerSubType(typeof(SelectQuestionRes), DiscriminatorValue = "select")]
public abstract class QuestionRes
{
    
}

public class ConfirmQuestionRes:QuestionRes
{
    public required string Message { get; set; }
    public string? Desc { get; set; }
    public string? Default { get; set; }
    public string? ErrorMessage { get; set; }
}

public class DateQuestionRes:QuestionRes
{
    public required string Message { get; set; }
    public string? Default { get; set; }
    public string? Desc { get; set; }
    public string? MinDate { get; set; }
    public string? MaxDate { get; set; }
}

public class CheckboxQuestionRes:QuestionRes
{
    public required string Message { get; set; }
    public string? Desc { get; set; }
    public required string[] Options { get; set; }
}

public class PasswordQuestionRes:QuestionRes
{
    public required string Message { get; set; }
    public string? Desc { get; set; }
    public bool? Confirmation { get; set; }
}

public class SelectQuestionRes:QuestionRes
{
    public required string Message { get; set; }
    public required string[] Options { get; set; }
    
    public string? Desc { get; set; }
}

public class TextQuestionRes:QuestionRes
{
    public required string Message { get; set;  }
    
    public string? Default { get; set;  }
    public string? Desc { get; set;  }
    public string? Initial { get; set;  }
}

File: sdks/node/src/api/core/question_res.ts
interface ConfirmQuestionRes {
  message: string;
  id: string;
  desc: string | null;
  default: boolean | null;
  errorMessage: string | null;
  type: 'confirm'; // Type discriminator value
}

interface DateQuestionRes {
  message: string;
  id: string;
  default: string | null;
  desc: string | null;
  minDate: string | null;
  maxDate: string | null;
  type: 'date'; // Type discriminator value
}

interface CheckboxQuestionRes {
  message: string;
  id: string;
  desc: string | null;
  options: string[];
  type: 'checkbox'; // Type discriminator value
}

interface PasswordQuestionRes {
  message: string;
  id: string;
  desc: string | null;
  confirmation: boolean | null;
  type: 'password'; // Type discriminator value
}

interface SelectQuestionRes {
  message: string;
  id: string;
  options: string[];
  desc: string | null;
  type: 'select'; // Type discriminator value
}

interface TextQuestionRes {
  message: string;
  id: string;
  default: string | null;
  desc: string | null;
  initial: string | null;
  type: 'text'; // Type discriminator value
}

type QuestionRes =
  | ConfirmQuestionRes
  | DateQuestionRes
  | CheckboxQuestionRes
  | PasswordQuestionRes
  | SelectQuestionRes
  | TextQuestionRes;

// export all
export type {
  QuestionRes,
  ConfirmQuestionRes,
  DateQuestionRes,
  CheckboxQuestionRes,
  PasswordQuestionRes,
  SelectQuestionRes,
  TextQuestionRes,
};
